### ISSUE TITLE
"Go to Test"/"Run Test at Cursor" and editor decorations are not kept in-sync

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#133441]
This might be by design, but it feels a bit awkward (and if intended, I think should be noted [in the test runner docs](https://code.visualstudio.com/api/extension-guides/testing)).

If an extension produces `Range` for a test, it is used for three things:

1. The Go to Test command on the test explorer context menu
2. The Run Test at Cursor command
3. The editor decorations for running tests

The button one updates automatically as the user types in the document (as do other decorations), but the first two do not. This means that "Go to Test" and "Run Test at Cursor" can do the wrong thing even when the editor clearly shows the decorations in the right places. Here, Go-to-Test goes to the original location (before I inserted a load of newlines), even though the decorations are all good.

![Sep-20-2021 13-09-02](https://user-images.githubusercontent.com/1078012/134000165-00cf58c5-059d-4331-9ecb-6fe4570ca859.gif)

It seems like this should be consistent. Otherwise, extensions need to be aware that they need to live-update the `Range` on each test node as the document changes to keep things working as users expect (this is probably required for things to work across reloads etc., but if it should at least be documented).

**Comment by connor4312** [comment]
> extensions need to be aware that they need to live-update the Range on each test node as the document changes to keep things working as users expect

This is suggested for extensions that can support it. It's alluded to in the code sample here https://code.visualstudio.com/api/extension-guides/testing#discovering-tests

**Comment by DanTup** [comment]
Yeah, I think it's very easy to miss that. I think it would be useful to make clearer the implications of not doing this. I actually wondered if this would happen previously, and I used the decorations as confirmation that VS Code was keep things up-to-date for me - then I discovered it only did for the decorations.

**Comment by connor4312** [comment]
I can see if we can do the same reconciliation, but this might not be feasible

**Comment by DanTup** [comment]
Yeah, I thought it might be hard so I've started working on keeping it updated from the language server. I think it's fine as long as it's a little clearer. Slightly related, https://github.com/microsoft/vscode/issues/54147 could be useful for things like this for languages/frameworks that don't have a good way to update the info from the document buffer changes.

**Comment by connor4312** [comment]
Ah, I see, Markers are what I'm after. Thanks, I should be able to manage it from here 🙂
